refactor: add migration guide (#611)
asyncapi/parser-js: docs/migration-guide-v2.md

> # Parser v1 to v2 Migration Guide

Reviewer: Maybe it makes sense to start of with "why v2" 🤔

Author: Overall, the first point of the General indicates why v2 - because a lot of breaking changes. Do we really need some description for that?

Reviewer: I guess not no 😄

## General

> 

Reviewer: I am missing a bit of context why I am reading a list here.

Author: Maybe we will change that header to the `tl;dr`? 😆 Because it's for that.

Reviewer: Would make sense yes 😄

# Parser v1 to v2 Migration Guide

> - The document's AsyncAPI API has been rewritten from old to new supporting the [Intent API](https://github.com/asyncapi/parser-api) - there is a function that converts the new API to the old one for backward compatibility. 

Reviewer 2: What about this suggestion, @magicmatatjahu? I think it improves the readability of the first point.

> - The document's AsyncAPI API has been rewritten from old to new supporting the [Intent API](https://github.com/asyncapi/parser-api) - there is a function that converts the new API to the old one for backward compatibility. 

Author: I forgot about this 😄 Applied.

- Package source code has been rewritten to the TypeScript alongside with rewritten from CJS (CommonJS) to ESM (EcmaScript) modules.
- The functionality of the Parser was wrapped with the class. This means that from now on it's possible to have a dozen other Parser instances with different configurations.
- Validation is powered by [Spectral](https://github.com/stoplightio/spectral). This allows better validation, checking not only syntax errors, but also informing about good practices, such as defining `operationID` field etc. In addition, Spectral uses a different references resolver, which causes changes in defining custom resolvers.
- AsyncAPI document validation (without parsing) is available via a separate method - `.validate()`.
- The function `.parse()` returns parsed AsyncAPi `document` and possible `diagnostics` based on validation configuration.  
- The function `parseFromUrl()` has been deprecated in favour of `fromURL` standalone function.
- The Custom Schema Parser interface has been changed with an additional function `validate()`.

> - Stringify and unstringify the parsed document is now a separate functionality.

Author: Stringify and unstringify functionalities aren't related to the ref resolvers.

- Stringify and unstringify the parsed document is now a separate functionality.

> - Package is fully [tree-shakable](https://developer.mozilla.org/en-US/docs/Glossary/Tree_shaking).

Reviewer 2: ```suggestion
```

This is already described in the Typescript and ESM section.

Author: Just because some lib is written with ESM does not mean it is fully tree-shakable :) it is better to list it here so that people do not have to look in other subsections.

Author: Ok, I changed second point in the list to the:

> Source code has been rewritten in TypeScript and ESM. This makes the package fully [tree-shakable](https://developer.mozilla.org/en-US/docs/Glossary/Tree_shaking). [Read more](#typescript-and-esm).

What do you think?

Reviewer 2: Yeah, that's much better. Thanks!

### New Intent API

The old Document API based 1:1 on AsyncAPI Specification has been rewritten to support the new so-called [Intent API](https://github.com/asyncapi/parser-api), which makes it easier to operate on the document and retrieve collections of needed objects from the document. Due to the fact that the new API does not stick firmly to AsyncAPI Specification, it allows for better integration with future versions of AsyncAPI. User does not need to know the new API against the new version of AsyncAPI - all methods will remain the same with the inclusion of new/deprecated fields. Also new API has been enhanced with some additional functionality like retrieving the JSON Path of the object using the `.jsonPath()` method etc.

> ### EcmaScript Modules

Reviewer 2: ```suggestion
### TypeScript and ESM
```

> The source code of the package was rewritten from CJS (CommonJS) to ESM (EcmaScript) modules. This allows the fully [tree-shaking](https://developer.mozilla.org/en-US/docs/Glossary/Tree_shaking) and much better integration in browser-type environments like SPA applications. In addition, the package is published with:

Reviewer 2: ```suggestion
The source code of the package has been rewritten to ESM (EcmaScript modules). This enables [tree-shaking](https://developer.mozilla.org/en-US/docs/Glossary/Tree_shaking) and much better integration in browser-based environments. In addition, the package exposes a `cjs` directory containing the CommonJS version of the parser. For compatibility with older environments.
```

> - `esm` (ESM modules) folder mainly for browser applications and newer versions of NodeJS, to support fully [tree-shaking](https://developer.mozilla.org/en-US/docs/Glossary/Tree_shaking) 
> - `cjs` (CJS modules) folder for old versions of NodeJS

Reviewer 2: You don't need that now :)

```suggestion
```

Author: Aaa ok, I read another comment.

import { Parser } from '@asyncapi/parser';
import myCustomSchemaParser from './my-custom-schema-parser';

const parser = new Parser();
parser.registerSchemaParser(myCustomSchemaParser);
```

### Standalone stringify and unstringify functionality

The ability to stringify and unstringify (parsing of stringified AsyncAPI document) was possible as methods in the old `AsyncAPiDocument` instance. Currently, both functions are standalone:

```ts
import { Parser, stringify, unstringify } from '@asyncapi/parser';

const parser = new Parser();

const { document } = await parser.parse(`
  asyncapi: '2.4.0'
  info:
    title: Example AsyncAPI specification
    version: '0.1.0'
  channels:
    example-channel:
      subscribe:
        message:
          payload:
            type: object
            properties:
              exampleField:
                type: string
              exampleNumber:
                type: number
              exampleDate:
                type: string
                format: date-time
`);

if (document) {
  // stringify function returns string type
  const stringifiedDocument = stringify(document);
  // unstringify function returns new AsyncAPIDocument instance

>   const unstringifiedDocument = unstringify(stringifiedDocument);

Reviewer 2: Wait 🤔 Isn't `unstringify` just the `.parse()` method? Why is it provided as a separate function?

And why isn't `.stringify()` a method of the parsed document instead? I mean, why not `document.stringify()`?

Author: > Wait 🤔 Isn't unstringify just the .parse() method? Why is it provided as a separate function?

If you put stringified document to the `.parse()` method, you will fire `unstringify()` function, don't worry, but then you need bundle all code related to the Parser, even when you don't need that, e.g. our html-template - `unstringify` function is for such a situation, to reduce bundled size of app.

> And why isn't .stringify() a method of the parsed document instead? I mean, why not document.stringify()?

To treeshake not needed functionality as much as possibility. If you will add `stringify` method to the AsyncAPIDocument class then this code will be always bundled in final code (even if you won't use that method).

Reviewer 2: Alright 👍 That makes sense. Thanks for the explanation!

`);

if (document) {
  // stringify function returns string type
  const stringifiedDocument = stringify(document);
  // unstringify function returns new AsyncAPIDocument instance
  const unstringifiedDocument = unstringify(stringifiedDocument);
}
```

### Convert new API to the old one

Due to the fact that a large part of the AsyncAPI tooling ecosystem uses a Parser with the old API and rewriting the tool for the new one can be time-consuming and difficult, the package exposes the `convertToOldAPI()` function to convert new API to old one:

```js
import { Parser, convertToOldAPI } from '@asyncapi/parser';

const parser = new Parser();
const { document } = parser.parse(...);
const oldAsyncAPIDocument = convertToOldAPI(document);
```

> **Warning**
> The old api will be supported only for a certain period of time. The target date for turning off support of the old API is around the end of January 2023.

If there is a need to convert an existing instance of the old API to the new one, the library provides the `convertToNewAPI` function:

```js
import { Parser, convertToNewAPI } from '@asyncapi/parser';

const newAsyncAPIDocument = convertToNewAPI(oldDocument);
```

> **Warning**

> > Due to the fact that the new Parser validates the docuemnt more restrictively, it may be that a parsed document in the old way will not work 100% in accordance with the new API.

Reviewer 2: ```suggestion
> Due to the fact that the new Parser validates the document more restrictively, it may be that a parsed document in the old way will not work 100% in accordance with the new API.
```
